OpenSim/Tests/Wrapping/testWrapCylinder.cpp

```diff
@@ -0,0 +1,314 @@
+        model.setUseVisualizer(visualize);
+
+        const SimTK::State& state = model.initSystem();
+
+        if (visualize) {
+            model.realizeVelocity(state);
+            model.getVisualizer().show(state);
+        }
+
+        model.updComponent<PathSpring>("spring").getLength(state);
+        WrapResult wrapResult = model.updComponent<PathSpring>("spring")
+                .getGeometryPath()
+                .getWrapSet()
+                .get("pathwrap")
+                .getPreviousWrap();
+
+        // Convert the WrapResult to a WrapTestResult for convenient assertions.
+        WrapTestResult result;
+        result.path = PathSegment{
+            wrapResult.r1,
+            wrapResult.r2,
+        };
+        result.length = wrapResult.wrap_path_length;
+        result.noWrap = wrapResult.wrap_pts.size() == 0;
+
+        return result;
+    }
+
+    // Simulates and tests a wrapping case:
+    // - Computes the wrapping result given the input.
+    // - Tests if computed result is equal to the expected result.
+    // - Returns a string containing info on the failed test, or an empty string
+    // if succesful.
+    std::string TestWrapping(
+        const WrapInput& input,
+        const WrapTestResult& expected,
+        WrappingTolerances tol,
+        std::string testCase,
+        bool visualize = false)
+    {
+
+        WrapTestResult result = solve(input, visualize);
+
+        if (!IsEqual(result, expected, tol.position)) {
+            std::ostringstream oss;
+            oss << "\nFAILED: case = " << testCase;
+            oss << "\n    input = " << input;
+            oss << "\n    result = " << result;
+            oss << "\n    expected = " << expected;
+            oss << "\n    caused by: "
+                << "Expected and simulated result not equal to within tolerance = "
+                << tol.position;
+            return oss.str();
+        }
+
+        return std::string{};
+    }
+
+}
+
+int main()
+{
+    std::string failLog;
+    WrappingTolerances tolerance;
+
+    WrapInput input {};
+    input.radius = 1.;
+    input.cylinderLength = 10.;
+
+    // =========================================================================
+    // ======================== No-Wrapping Case ===============================
+    // =========================================================================
+
+    // Placing the path inside the cylinder results in a no-wrap.
+    std::string name = "Inside cylinder";
+    input.path = {{}, {}};
+
+    WrapTestResult expected = WrapTestResult::NoWrap();
+
+    failLog += TestWrapping(input, expected, tolerance, name);
+
+    // =========================================================================
+    // ======================= Perpendicular Case ==============================
+    // =========================================================================
+
+    name = "Perpendicular";
+    input.path = {{2, -2, 0}, {-2, 2.1, 0}};
+
+    expected.path = {
```

Comment:
Where are these numbers from?

Reply:
My idea was to add more assertions for each case, such that (hopefully) all edge cases and interesting wrapping conditions can be captured. But before adding those cases, I would like to add the testing of certain geometric properties, such as path gradient, wrapping direction etc. So I just added a simple case for which the wrapping path is very accurate, but is otherwise not necessarily interesting.

But yeah, at this moment, there is no way of testing if this wrapping is accurate, so it seems a bit random.

Reply:
A comment along the lines of "these were scooped from OpenSim 4.5", "calculated by hand", etc. is enough here